### ISSUE TITLE
Incorrect metrics for ragged tensor as output

### ISSUE THREAD

**Issue by HannaLochOlszewska** [tensorflow/tensorflow#51444]
**System information**
- OS Platform: Windows 10
- TensorFlow installed from pip, version 2.5.0
- Python version: 3.8

**Describe the current behavior**
The metrics for the ragged tensors are calculating incorrectly. 

**Standalone code to reproduce the issue**
The model:

```
class dummy(tf.keras.Model):
    def __init__(self):
        super(dummy, self).__init__()

    def call(self, inputs):
        lens = tf.map_fn(lambda x: len(x), inputs, fn_output_signature=tf.int32, name='get_lengths')
        tensored_input = inputs.to_tensor(0, shape=[None, 10])
        x = Lambda(lambda x: tf.cast(x, dtype=tf.float32))(tensored_input)
        outputs = tf.RaggedTensor.from_tensor(x, lengths=lens)
        return outputs
```

Running:
```
x_dataset = tf.data.Dataset.from_tensor_slices( tf.ragged.constant([[0,0,0,0,0],[0],[0,0,0],[0,0,0,0]]))
y_dataset = tf.data.Dataset.from_tensor_slices( tf.ragged.constant([[1,1,1,1,1],[1],[1,1,1],[1,1,1,1]]))
tensor_dataset = tf.data.Dataset.zip((x_dataset, y_dataset))
# Create batches
batched_ds = tensor_dataset.batch(2, drop_remainder=True)

model = dummy()
model.compile(loss='binary_crossentropy',metrics=['accuracy'])
history = model.fit(batched_ds,epochs=2)
```
Result (history.history) - 100% accuracy despite clear dismatches.

`
{'loss': [15.424947738647461, 15.424947738647461], 'accuracy': [1.0, 1.0]}
`

**Comment by sushreebarsa** [comment]
@HannaLochOlszewska In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by HannaLochOlszewska** [comment]
```
import tensorflow as tf
from tensorflow.keras.layers import Lambda

class dummy(tf.keras.Model):
    def __init__(self):
        super(dummy, self).__init__()

    def call(self, inputs):
        lens = tf.map_fn(lambda x: len(x), inputs, fn_output_signature=tf.int32, name='get_lengths')
        tensored_input = inputs.to_tensor(0, shape=[None, 10])
        x = Lambda(lambda x: tf.cast(x, dtype=tf.float32))(tensored_input)
        outputs = tf.RaggedTensor.from_tensor(x, lengths=lens)
        return outputs

x_dataset = tf.data.Dataset.from_tensor_slices( tf.ragged.constant([[0,0,0,0,0],[0],[0,0,0],[0,0,0,0]]))
y_dataset = tf.data.Dataset.from_tensor_slices( tf.ragged.constant([[1,1,1,1,1],[1],[1,1,1],[1,1,1,1]]))
tensor_dataset = tf.data.Dataset.zip((x_dataset, y_dataset))
# Create batches
batched_ds = tensor_dataset.batch(2, drop_remainder=True)

model = dummy()
model.compile(loss='binary_crossentropy',metrics=['accuracy'])
history = model.fit(batched_ds,epochs=2)
# Accuracy should be 0 here obviously and it's not
print(history.history)
```

**Comment by sushreebarsa** [comment]
@HannaLochOlszewska Could you please take a look on the[ link1](https://www.tensorflow.org/guide/ragged_tensor) , [link2](https://www.tensorflow.org/api_docs/python/tf/RaggedTensor?version=nightly) and let us know if it helps ? Thank you!

**Comment by HannaLochOlszewska** [comment]
Hi @sushreebarsa, the links didn't help much. But I managed to find out that 
`model.compile(loss='binary_crossentropy',metrics=['accuracy'])` yields incorrect accuracy results, while
`model.compile(loss='binary_crossentropy',metrics=[tf.keras.metrics.Accuracy()])` seems to work fine.

**Comment by sushreebarsa** [comment]
@HannaLochOlszewska Thank you for the update! I also tried to reproduce the same in colab using TF v2.5, please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/2200d5d521619776ea308a12eb82b558/51444.ipynb) and confirm the same .Kindly move this issue to closed status if it is resolved for you ? Thanks!

**Comment by HannaLochOlszewska** [comment]
Yes, indeed, it's resolved - thanks a lot for your assistance!

For the future reference - the problem was only in the accuracy metric that was passed, not ragged tensors at all, my bad.
According to tf.keras.Model.compile() documentation:
>  When you pass the strings 'accuracy' or 'acc', we convert this to one of tf.keras.metrics.BinaryAccuracy, tf.keras.metrics.CategoricalAccuracy, tf.keras.metrics.SparseCategoricalAccuracy based on the loss function used and the model output shape. We do a similar conversion for the strings 'crossentropy' and 'ce' as well.

**Comment by sushreebarsa** [comment]
@HannaLochOlszewska Glad that it worked for you .Could you please close this issue as it is resolved ?Thank you!
